entrypoint/help_utils.go

```diff
@@ -0,0 +1,155 @@
+
+Commands:
+
+{{range .Commands}}{{if not .HideHelp}}   {{join .Names ", "}}{{ "\t"}}{{.Usage}}{{ "\n" }}{{end}}{{end}}{{end}}{{if .VisibleFlags}}{{end}}
+`
+
+const CLI_COMMAND_HELP_TEMPLATE = `Usage: {{if .UsageText}}{{.UsageText}}{{else}}{{.HelpName}}{{if .VisibleFlags}} [options]{{end}} {{if .ArgsUsage}}{{.ArgsUsage}}{{else}}[args]{{end}}{{end}}{{if .Description}}
+
+{{.Description}}{{end}}{{if .VisibleFlags}}
+
+Options:
+
+   {{range .VisibleFlags}}{{.}}
+   {{end}}{{end}}
+`
+
+// HelpPrinter that will wrap the text at
+// HELP_TEXT_LINE_WIDTH characters, while preserving
+// indentation and table tabs. Currently only works with
+// tables delimited by `\t` and with only 2 columns.
+func WrappedHelpPrinter(out io.Writer, templateString string, data interface{}) {
+	funcMap := template.FuncMap{
+		"join":                  strings.Join,
+		"PrefixedFirstFlagName": PrefixedFirstFlagName,
+	}
+
+	templ := template.Must(template.New("help").Funcs(funcMap).Parse(templateString))
+	rendered := bytes.NewBufferString("")
+	err := templ.Execute(rendered, data)
+	if err != nil {
+		return
+	}
+
+	writer := tabwriter.NewWriter(out, 0, 8, 2, ' ', 0)
+	for _, line := range strings.Split(rendered.String(), "\n") {
+		tab := HelpTableAwareDetermineIndent(line, "\t+")
+		wrappedLine := TabAwareWrapText(line, HELP_TEXT_LINE_WIDTH, tab)
+		fmt.Fprintln(writer, wrappedLine)
+	}
+	writer.Flush()
+}
+
+// Similar functionality to regexp.Split, but returns the delimited strings
+// with the trailing delimiter appended to it.
+// Example:
+//    re := regexp.MustCompile("\\s+")
+//    text := "one two three"
+//    out := SplitKeepDelimiter(re, text)
+//    out == ["one ", "two ", "three"]
+func SplitKeepDelimiter(re *regexp.Regexp, str string) (out []string) {
+	indexes := re.FindAllStringIndex(str, -1)
+	if indexes == nil {
+		return append(out, str)
+	}
+	cur := 0
+	for _, index := range indexes {
+		out = append(out, str[cur:index[1]])
+		cur = index[1]
+	}
+	if cur != len(str) {
+		out = append(out, str[cur:])
+	}
+	return out
+}
+
+// Wrap text to line width, while preserving any indentation
+func TabAwareWrapText(text string, lineWidth int, tab string) string {
+	wrapped := ""
+	re := regexp.MustCompile("\\s+")
+	words := SplitKeepDelimiter(re, text)
+	if len(words) == 0 {
+		return wrapped
+	}
+	// Keep on consuming words to current line until current line reaches
+	// lineWidth, at which point we start a new line. Keep in mind that word is
+	// word + whitespace to next word.
+	wrapped = words[0]
+	// NOTE: This tabLength is not exactly correct, as elastic
+	//       tabstops work with cells which may cause the tabs
+	//       to expand beyond 8 spaces. Nonetheless the goal
+	//       is to avoid overflowing beyond the lineWidth and
+	//       this should be a better approximation than 1
+	//       char.
+	tabLength := 8
+	currentLineLength := TabAwareStringLength(wrapped, tabLength)
+	for _, word := range words[1:] {
```

Comment:
In general, this sort of code can be easier to read and maintain if written recursively:

```go
func TabAwareWrapText(words []string, lineWidth int, tab string, outputLinesSoFar []string): string {
  if len(words) == 0 {
    return strings.Join(outputLinesSoFar, "\n")
  }

  head = words[0]
  tail = words[1:]

  if lineLengthExceeded(outputLinesSoFar, head, lineWidth) {
    return TabAwareWrapText(tail, lineWidth, tab, addWordOnNewLine(outputLinesSoFar, head))
  } else {
    return TabAwareWrapText(tail, lineWidth, tab, addWordToExistingLine(outputLinesSoFar, head))
  }
}
```

Not sure it's worth refactoring in this case, as (a) Go isn't exactly a functional programming language and (b) I'm not sure it does tail call optimization, which may be a problem with long strings.

Reply:
Agree. I was going to write recursively, but then I learned that [the maintainers of Go isn't too keen on supporting tail call optimization](https://groups.google.com/forum/#!msg/golang-nuts/0oIZPHhrDzY/2nCpUZDKZAAJ) so opted not to. Sad :(

Reply:
Gah. Stupid Go.